src/ipa_pc/mod.rs

```diff
@@ -456,35 +622,79 @@ impl<G: AffineCurve, D: Digest> PolynomialCommitment<G::ScalarField> for InnerPr
 
-            let randomness = if let Some(h) = hiding_bound {
-                Randomness::rand(h, degree_bound.is_some(), rng)
+            let key_len = ck.comm_key.len();
+            let p_len = polynomial.coeffs.len();
+            let segments_count = p_len / key_len + if p_len % key_len != 0 { 1 } else { 0 };
+
+            let randomness = if let Some(_) = hiding_bound {
+                Randomness::rand(segments_count, degree_bound.is_some(), rng)
             } else {
-                Randomness::empty()
+                Randomness::empty(segments_count)
             };
 
-            let comm = Self::cm_commit(
-                &ck.comm_key[..(polynomial.degree() + 1)],
-                &polynomial.coeffs,
-                Some(ck.s),
-                Some(randomness.rand),
-            )
-            .into_affine();
+            let comm: Vec<G>;
 
-            let shifted_comm = degree_bound.map(|d| {
-                Self::cm_commit(
-                    &ck.comm_key[(ck.supported_degree() - d)..],
-                    &polynomial.coeffs,
-                    Some(ck.s),
-                    randomness.shifted_rand,
-                )
-                .into_affine()
+            if segments_count > 1 {
+
+                // split poly in segments and commit all of them without shifting
+                comm = (0..segments_count).into_iter().map(
+                    |i| {
+                        Self::cm_commit(
+                            &ck.comm_key,
+                            &polynomial.coeffs[i * key_len..core::cmp::min((i + 1) * key_len, p_len)],
+                            Some(ck.s),
+                            Some(randomness.rand[i]),
+                        ).into_affine()
+                    }
+                ).collect();
+
+            } else {
+
+                comm = vec![
+                    Self::cm_commit(
+                        &ck.comm_key[..(polynomial.degree() + 1)],
+                        &polynomial.coeffs,
+                        Some(ck.s),
+                        Some(randomness.rand[0]),
+                    )
+                    .into_affine()
+                ];
+            }
+
+            // committing only last segment shifted to the right edge
+            let shifted_comm = degree_bound.and_then(|degree_bound| {
+                let degree_bound_len = degree_bound + 1; // Convert to the maximum number of coefficients
+                if degree_bound_len % key_len != 0 {
+                    if segments_count > 1 {
+                        Some(
+                            Self::cm_commit(
+                                &ck.comm_key[key_len - (degree_bound_len % key_len)..],
+                                &polynomial.coeffs[(segments_count - 1) * key_len..p_len],
+                                Some(ck.s),
+                                randomness.shifted_rand,
+                            ).into_affine()
+                        )
+                    } else {
```

Comment:
I think here the separation into the cases segments_count >= 1 is not arguable.

Reply:
Done